### PR TITLE
Do not add return dict items to callback_metrics

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Deletes legacy support for adding items to `callback_metrics` by returning them in a dictionary.

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/6623

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 1]
```diff
@@ -446,10 +427,9 @@ def log_train_epoch_end_metrics(
 
         # TODO: deprecate 1.0
         else:
-            out = self.__run_legacy_training_epoch_end(
-                num_optimizers, epoch_output, model, is_result_obj, epoch_callback_metrics
+            epoch_log_metrics, epoch_progress_bar_metrics = self.__run_legacy_training_epoch_end(
+                num_optimizers, epoch_output, model, is_result_obj
             )
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:432]: we file a follow up issue to fully remove this path

File: CHANGELOG.md [hunk 2]
```diff
@@ -147,6 +147,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Removed deprecated `LightningModule` `hparams` setter ([#6207](https://github.com/PyTorchLightning/pytorch-lightning/pull/6207))
 
 
+- Removed legacy code to include `step` dictionary returns in `callback_metrics` ([#6682](https://github.com/PyTorchLightning/pytorch-lightning/pull/6682))
+
+
```
Comment (ananthsub) [inline CHANGELOG.md:152]: Should we also mention using `self.log(...)` or `self.log_dict(...)` in place of returning a dictionary?

File: tests/trainer/logging_/test_logger_connector.py [hunk 3]
```diff
@@ -479,6 +479,27 @@ def test_step(self, *args, **kwargs):
...
+
+        def validation_step(self, batch, *args, **kwargs):
+            return {"val": torch.tensor([0, 1])}
+
+        def test_step(self, batch, *args, **kwargs):
+            return {"test": torch.tensor([0, 1])}
+
+    model = TestModel()
+    model.validation_epoch_end = None
+    model.test_epoch_end = None
```
Comment (ananthsub) [inline tests/trainer/logging_/test_logger_connector.py:495]: do we already have a test to ensure that the step outputs are still returned to the `_epoch_end` hooks?

**Review by ananthsub — APPROVED** [review]
great fix! awesome to see the trainer/logger connector being simplified

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:432]: No, but we need to clean this asap. I'll try to split it among several PRs as there is a lot of code to remove

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 4]
```diff
@@ -380,20 +365,16 @@ def __process_eval_epoch_end_results_and_log_legacy(self, eval_results):
                 eval_results = [eval_results]
 
             num_loaders: int = self.trainer.evaluation_loop.num_dataloaders
-            prog_bar_metrics, log_metrics, callback_metrics = {}, {}, {}
+            prog_bar_metrics, log_metrics = {}, {}
 
             for result_idx, result in enumerate(eval_results):
-                _, prog_bar_metrics, log_metrics, callback_metrics, _ = self.trainer.process_dict_result(result)
+                _, prog_bar_metrics, log_metrics, _ = self.trainer.process_dict_result(result)
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:371]: What are the extra returned objects ?

File: docs/source/ecosystem/asr_nlp_tts.rst [hunk 5]
```diff
@@ -270,12 +270,12 @@ with PyTorch Lightning since every NeMo model is a Lightning Module.
                 log_probs=log_probs, targets=transcript, input_lengths=encoded_len, target_lengths=transcript_len
             )
             wer_num, wer_denom = self._wer(predictions, transcript, transcript_len)
-            tensorboard_logs = {
+            self.log_dict({
```
Comment (tchaton) [inline docs/source/ecosystem/asr_nlp_tts.rst:273]: Nice !

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 6]
```diff
@@ -185,9 +183,6 @@ def cache_training_step_metrics(self, opt_closure_result):
             batch_log_metrics = opt_closure_result.training_step_output.log_metrics
             logged_metrics_tmp.update(batch_log_metrics)
 
-            callback_metrics = opt_closure_result.training_step_output.callback_metrics
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:188]: Why can we remove this ?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 7]
```diff
@@ -380,20 +365,16 @@ def __process_eval_epoch_end_results_and_log_legacy(self, eval_results):
...
 
             for result_idx, result in enumerate(eval_results):
-                _, prog_bar_metrics, log_metrics, callback_metrics, _ = self.trainer.process_dict_result(result)
+                _, prog_bar_metrics, log_metrics, _ = self.trainer.process_dict_result(result)
 
                 if num_loaders > 1:
-                    self.__process_eval_epoch_end_results_and_log_legacy_update(
-                        prog_bar_metrics, log_metrics, callback_metrics
-                    )
+                    self.__process_eval_epoch_end_results_and_log_legacy_update(prog_bar_metrics, log_metrics)
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:374]: Mind clean this `num_loaders > 1` and `num_loaders == 1`

File: tests/trainer/logging_/test_logger_connector.py (hunk 3 above)
Comment (tchaton) [reply tests/trainer/logging_/test_logger_connector.py:495]: Can this test be extended to make sure of it ?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:371]: `output, progress_bar_metrics, log_metrics, hiddens`

To be removed in separate PRs

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 6 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:188]: Because `training_step_output.callback_metrics` was removed (is now empty)

**Review by tchaton — APPROVED** [review]
Great work ! That's a very important step toward cleaning all this mess :)

**Review by justusschock — APPROVED** [review]
